File: Python3/Medium/Accepted/circular-array-loop_6.py
class Solution:
    def circularArrayLoop(self, nums: List[int]) -> bool:
        
        n = len(nums)
        for i in range(n):
            if nums[i] == 0:
                continue
            slow = i
            fast = (i + nums[i]) % n
            if nums[fast] * nums[i] < 0:
                continue
            while nums[fast] * nums[i] > 0 and nums[(fast + nums[fast]) % n] * nums[i] > 0:
                if slow == fast:
                    if slow == (slow + nums[slow]) % n:
                        break
                    return True
                slow = (slow + nums[slow]) % n
                fast = (fast + nums[fast]) % n
                fast = (fast + nums[fast]) % n
            slow = i
            val = nums[i]
            while nums[slow] * val > 0:
                tmp = slow
                slow = (slow + nums[slow]) % n
                nums[tmp] = 0
        return False

File: Rust/Medium/TimeLimitExceeded/circular-array-loop_6.rs
impl Solution {
    pub fn circular_array_loop(nums: Vec<i32>) -> bool {
        

        let mut nums = nums;
        let mut i = 0;
        while i < nums.len() {
            let mut j = i;
            let mut k = i;
            while nums[j] * nums[(j + nums[j] as usize) % nums.len()] > 0
                && nums[j] * nums[(k + nums[k] as usize) % nums.len()] > 0
                && nums[j] * nums[(k + nums[k] as usize * 2) % nums.len()] > 0
            {
                j = (j + nums[j] as usize) % nums.len();
                k = (k + nums[k] as usize * 2) % nums.len();
                if j == k {
                    if j == (j + nums[j] as usize) % nums.len() {
                        break;
                    }
                    return true;
                }
            }
            let val = nums[i];
            while nums[i] * val > 0 {
                let next = (i + nums[i] as usize) % nums.len();
                nums[i] = 0;
                i = next;
            }
        }
        false
    }
}